### PR TITLE
Knife should give a useful error when the chef_server_url isn't a chef server API

### PR DESCRIPTION
**Pull request by jeunito**
### Description
Raises an exception if the value of chef_server_url is not a valid chef_server_api

### Issues Resolved

https://github.com/chef/chef/issues/5552

### Check List

- [x] New functionality includes tests
- [x] All tests pass
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

**Comment by stevendanna** [comment]
Do we know what this query will return on an open source chef 11 server?

**Comment by jeunito** [comment]
@stevendanna no I did not consider that. Should it be supported too?

**Comment by stevendanna** [comment]
Unfortunately, we don't have great numbers on the deployed versions of Chef Server out there.  I'd definitely prefer if there was a way to do this without breaking Chef Server 11 if it isn't too onerous.

**Comment by jeunito** [comment]
@coderanger I just assumed that the value of ```chef_server_url``` is always of the form 
> http://fqdn/organizations/NAME

which is what I see on a hosted chef setup. 

Hence, it is actually sending a request to https://docs.chef.io/api_chef_server.html#organizations-name. 

Could this be different in an on-premise setup?

**Comment by coderanger** [comment]
@jeunito That's part of the global API though, which we don't use for anything else in chef-client that I know of. There is no definition for a `/` call in https://docs.chef.io/api_chef_server.html#organization-endpoints

**Comment by coderanger** [comment]
@jeunito Basically we don't use that API from chef-client, it's part of the Erchef management API but not the "Chef Server API" which is a subset/different thing.

**Comment by coderanger** [comment]
Otherwise, for example, this would fail on Chef Server 11.

**Comment by stevendanna** [comment]
Because of the array of server implementations we might be talking to, an active check might be difficult.

I think my preferred alternate approach would be to avoid doing an extra request completely and add some heuristics to the error handling for failed requests to add a warning if the response looks like it might not be from a chef server.

That is, if you tried to list nodes but you got a 404 or some other error response, we look for things like whether the response was HTML rather than JSON and then print out a "This response does not
look like a response from a valid Chef Server, is your chef_server_url correct?" message.

## On GET /organizations/ORGNAME

I'm not sure I agree that this endpoint isn't part of the Chef API. Namely, it is

- Documented here: https://chef-server-api-docs.chef.io/#organizations__organization__get
- Supported by chef-zero
- Tested in oc-chef-pedant

However, the fact that it is a question at all likely means we should not add it as a dependency to every request. We've certainly treated it as an internal API as the contents have changed over time. Further, I could imagine situations were odd permissions issues could get in the way of an otherwise successful request.

## A possible active check
Unfortunately, supporting chef server 11, chef server 12, chef-zero, and goiardi might be difficult.

Perhaps we could do the following:

- GET to /clients/CLIENT_NAME where CLIENT_NAME is name of the
  client we are trying to make a request as.

- In the case of a 200, we can confirm that we got a client back.

- In the case of a 404, a compliant Chef server should still be
  serving up JSON content. If we see anything else, we can
  complain that it might not be a Chef Server.

- In the case of a 401 or 403, again we would likely have to
  check that the shape of the response looks correct.

- In the case of 50X we might want to "fail open" in case it was
  a transient failure.

## Some other thoughts

I originally thought we could use a response header. Unfortunately I couldn't not find one that is supported across implementation. Chef Server will return `x-ops-api-info` and `x-ops-server-api-version` but the former doesn't appears unsupported in chef-zero or goiardi and the later was only added in Chef Server 12.something.

We could use an endpoint that is more or less guaranteed to exist (such as `/clients`) but we would still have to have handling for authn and authz errors and this could be problematic for organizations where the return body will be large.

**Comment by jeunito** [comment]
Thanks Steven. I ll look into your suggestion.

**Comment by jeunito** [comment]
> In the case of a 404...If we see anything else, we can complain that it might not be a Chef Server.

@stevendanna unfortunately we can't inspect the content when we get a 404 due to https://github.com/chef/chef/blob/master/lib/chef/http.rb#L155-L163

How about just letting it fail for all of 401, 404, 403 and 5xx cases you mentioned above and putting something like the following: 

> There was an issue while trying to validate your chef_server_url. Please check that: 
> 1. The value of chef_server_url points to a chef server
> 2. This client or your user name has the appropriate permissions 
> 3. The client or user ```node or client name``` exists 

This would be one step better than just letting chef fail with an unclear error and no clues for the user as mentioned in https://github.com/chef/chef/issues/5552

**Comment by jeunito** [comment]
@stevendanna nvm the comment at https://github.com/chef/chef/pull/6253#issuecomment-316843590. I think I have found an answer to the initial perceived obstacle I mentioned.

**Comment by lamont-granquist** [comment]
Unfortunately I'm 100% opposed to anything which incurs an extra round trip and API call just to ping the server every time a ServerAPI object is created.  Creates more round trip delays and more load on the server.  I'd prefer to leave the problem unfixed than to fix it this way.

You could do something better with error decoration where when get raised errors like ECONNREFUSED or 404'd then ping to check if it looked like it was a chef server at all -- *after* you've already hit an error.  Ideally memoize this result so that you don't need to ping check for multiple broken requests (and maybe if you get any success set it so that it doesn't have to ping it in the future).

**Comment by lamont-granquist** [comment]
^ there's going to be an obviously annoying aspect to that which is that if you rescue in get() and then call get() you're getting recursive behavior so that needs to be handled carefully so that you don't cause a ping flood / stack size overflow

**Comment by stevendanna** [comment]
> You could do something better with error decoration where when get raised errors like ECONNREFUSED or 404'd then ping to check if it looked like it was a chef server at all -- after you've already hit an error. Ideally memoize this result so that you don't need to ping check for multiple broken requests (and maybe if you get any success set it so that it doesn't have to ping it in the future).

Yes, this is what I would like to see as well.  Heuristics on the /response/ to a request rather than a proactive check.

**Comment by jeunito** [comment]
@stevendanna @lamont-granquist ok thanks both. Will do that instead

**Comment by jeunito** [comment]
> ...ping to check if it looked like it was a chef server at all -- after you've already hit an error.

@stevendanna @lamont-granquist please see this [commit](https://github.com/chef/chef/pull/6253/commits/b3eec2dc4c27bba5256b1ae42edb70cf422997ac) that reflects this proposal

File: lib/chef/knife/list.rb [hunk 1]
```diff
@@ -104,6 +104,18 @@ def add_dir_result(result)
         begin
           children = result.children.sort_by { |child| child.name }
         rescue Chef::ChefFS::FileSystem::NotFoundError => e
+          begin
+            rest.get("/clients/#{Chef::Config[:client_name]}")
+          rescue Net::HTTPServerException => e
+            if e.response.class == Net::HTTPNotFound
+              require "json"
+              begin
+                JSON.load(e.response.body())
```
Comment (lamont-granquist) [inline lib/chef/knife/list.rb]: that should use ffi-yajl, and the require should get moved to the top of the file.

**Comment by lamont-granquist** [comment]
could probably use a rebase+squash of commits (and removal of the merge commit)

File: lib/chef/knife/list.rb (hunk 1 above)
Comment (thommay) [reply lib/chef/knife/list.rb]: This should happen in `chef/http` and not in individual classes.

Comment (jeunito) [reply lib/chef/knife/list.rb]: @thommay Did you mean ```chef/server_api```? I reckon that ```chef/http``` could be used somewhere else and this logic is specific to a chef server.

**Comment by jeunito** [comment]
@thommay @lamont-granquist please see the new commits. I have extracted the logic for the heuristic to server_api

**Pull request merged**